internal/client.go

```diff
@@ -0,0 +1,304 @@
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("request failed with code: %d", resp.StatusCode)
+	}
+
+	// Parse the response body into a BucketMetadata struct
+	var metadata types.BucketMetadata
+	err = json.NewDecoder(resp.Body).Decode(&metadata)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read bucket metadata: %v", err)
+	}
+
+	return &metadata, nil
+}
+
+func (c *Client) doRequestWithRetry(req *http.Request) (*http.Response, error) {
+	maxRetries := 5
+	backoffDelay := 3 * time.Second
+	maxBackoffDelay := 60 * time.Second
+
+	var resp *http.Response
+	var err error
+
+	for i := 0; i < maxRetries; i++ {
+		// Clone the request to avoid issues with mutated request objects
+		clonedReq, err := cloneRequest(req)
+		if err != nil {
+			return nil, fmt.Errorf("failed to clone request: %v", err)
+		}
+
+		resp, err = c.doSignedRequest(clonedReq)
+		if err != nil {
+			return nil, fmt.Errorf("failed to send request: %v", err)
+		}
+
+		// Check if the response status code indicates a server-side error (5xx)
+		if resp.StatusCode >= 500 {
+			resp.Body.Close()
+
+			// Exponential backoff before retrying
+			time.Sleep(backoffDelay)
+			backoffDelay *= 2 // Double the delay for each retry
+			if backoffDelay > maxBackoffDelay {
+				backoffDelay = maxBackoffDelay
+			}
+
+			continue
+		}
+
+		// Break out of the loop if the request was successful
+		break
+	}
+
+	return resp, err
+}
+
+func (c *Client) doSignedRequest(req *http.Request) (*http.Response, error) {
+	// Sign the request
+	err := c.signRequest(req)
+	if err != nil {
+		return nil, fmt.Errorf("failed to sign request: %v", err)
+	}
+
+	// Send the signed request using the wrapped http.Client
+	return c.httpClient.Do(req)
+}
+
+func (c *Client) bucketURL(bucketName string, queryParams map[string]string) string {
+	baseURL := fmt.Sprintf("%s/%s", c.endpoint, bucketName)
+	if len(queryParams) == 0 {
+		return baseURL
+	}
+
+	// Add query parameters to the URL
+	query := url.Values{}
+	for key, value := range queryParams {
+		query.Add(key, value)
+	}
+
+	return fmt.Sprintf("%s?%s", baseURL, query.Encode())
+}
+
+func (c *Client) signRequest(req *http.Request) error {
+	// Get the current time for the request
+	now := time.Now()
+
+	// Set default headers
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("Accept", "application/json")
+
+	// Buffer the request body if it exists
+	var bodyBytes []byte
+	if req.Body != nil {
+		var err error
+		bodyBytes, err = io.ReadAll(req.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read request body: %v", err)
+		}
+		req.Body = io.NopCloser(bytes.NewReader(bodyBytes))
+	}
+
+	// Determine the payload hash (empty string if no payload)
+	payloadHash := hex.EncodeToString(sha256.New().Sum(bodyBytes))
+
+	// Sign the request using the signer
+	err := c.signer.SignHTTP(context.TODO(), c.credentials, req, payloadHash, "s3", c.cfg.Region, now)
```

Comment:
region for us is always going to be auto. maybe hardcode it?

Reply:
Good point. Let me make the change.